### PR TITLE
Another step towards removing boost

### PR DESCRIPTION
**Pull request by emjotde**
* Adds Pathie as a replacement for boost::filesystem
* Adds zstr as a replacement for boost::filtering_stream

### CODE REVIEW

File: src/common/config_validator.cpp [hunk 1]
```diff
@@ -84,9 +84,6 @@ void ConfigValidator::validateOptionsTraining() const {
   ABORT_IF(!modelDir.empty() && !filesystem::isDirectory(modelDir),
            "Model directory does not exist");
 
-  ABORT_IF(!modelDir.empty() && !filesystem::canWrite(modelDir),
-           "No write permission in model directory");
```
Comment (snukky) [inline src/common/config_validator.cpp]: This will very likely break one of our regression tests, so we will need to update it.

File: src/common/logging.cpp [hunk 2]
```diff
@@ -150,9 +150,11 @@ void switchtoMultinodeLogging(std::string nodeIdStr) {
 
 
 namespace marian {
-  void noinline logCallStack(size_t skipLevels)
-  {
-    auto callStack = ::Microsoft::MSR::CNTK::DebugUtil::GetCallStack(skipLevels + 2, /*makeFunctionNamesStandOut=*/true);
-    checkedLog("general", "critical", "Call stack:{}", callStack);
+  std::string getCallStack(size_t skipLevels) {
+    return ::Microsoft::MSR::CNTK::DebugUtil::GetCallStack(skipLevels + 2, /*makeFunctionNamesStandOut=*/true);
```
Comment (snukky) [inline src/common/logging.cpp]: Does this require some external dependency or should be used only with a specific compilation settings? Regardless of whether it does or not, a comment might be helpful here.

File: src/common/logging.h [hunk 3]
```diff
@@ -46,13 +47,19 @@ namespace marian {
...
+    if(logger == nullptr)                                                      \
+      logger = createStderrLogger("general", "[%Y-%m-%d %T] Error: %v");       \
+    else                                                                       \
+      logger->set_pattern("[%Y-%m-%d %T] Error: %v");                          \
+    checkedLog("general", "critical", __VA_ARGS__);                            \
+    checkedLog("general", "critical", "Aborted from {} in {}:{}",              \
+               FUNCTION_NAME, __FILE__, __LINE__);                             \
+    logger->set_pattern("%v");                                                 \
+    checkedLog("general", "critical", marian::getCallStack(/*skipLevels=*/0)); \
+    std::abort();                                                              \
```
Comment (snukky) [inline src/common/logging.h]: This refactorization is not related to removing Boost dependency, right? Just a question.

File: src/layers/word2vec_reader.h [hunk 4]
```diff
@@ -18,8 +18,6 @@ class Word2VecReader {
     LOG(info, "[data] Loading embedding vectors from {}", fileName);
 
     io::InputFileStream embFile(fileName);
-    ABORT_IF(!embFile.isOpen(),
-             "Unable to open file with embeddings: " + fileName);
```
Comment (snukky) [inline src/layers/word2vec_reader.h]: We don't need this or the condition is checked implicitly by InputFileStream?

File: src/data/default_vocab.cpp [hunk 5]
```diff
@@ -199,10 +199,6 @@ class DefaultVocab : public VocabBase {
             "Specified vocab directory {} does not exist",
             dir.string());
 
-    ABORT_IF(!dir.empty() && !filesystem::canWrite(dir),
-            "No write permission in vocab directory {}",
-            dir.string());
```
Comment (snukky) [inline src/data/default_vocab.cpp]: As above, we need to remember that this might need updating one of the regression tests.

File: src/common/logging.h (hunk 3 above)
Comment (emjotde) [reply src/common/logging.h]: No, this was a general aesthetic thing.

File: src/common/logging.cpp (hunk 2 above)
Comment (emjotde) [reply src/common/logging.cpp]: There are no dependencies here. Just a weird cascade of namespaces.

File: src/data/default_vocab.cpp (hunk 5 above)
Comment (emjotde) [reply src/data/default_vocab.cpp]: I think our regression tests will remind us :)

File: src/layers/word2vec_reader.h (hunk 4 above)
Comment (emjotde) [reply src/layers/word2vec_reader.h]: We are now aborting on badbit and failbit. This was not working before. So a failure to read aborts anyway.
